File: AMER Semifinal 2008/D.py
#!/usr/bin/env python
#
# Problem: King
# Language: Python
# Author: KirarinSnow
# Usage: python thisfile.py <input.in >output.out
# Comments: Adapted from reference solution. Times out on large.


def compute():
    def solve(x, y, b):
        if x == c:
            x = 0
            y += 1
            if y == r: return 0
    
        if m[x][y][b] != -1:
            return m[x][y][b]

        bb = (b<<1) & ((1<<(c+1))-1)
    
        if a[y][x] != '.':
            m[x][y][b] = solve(x+1, y, bb)
        else:
            m[x][y][b] = solve(x+1, y, bb+1)
            if x and b&1:
                m[x][y][b] = max(m[x][y][b], 1+solve(x+1, y, bb-2))
            if x and b&(1<<c):
                m[x][y][b] = max(m[x][y][b], 1+solve(x+1, y, bb))
            if b&(1<<(c-1)):
                m[x][y][b] = max(m[x][y][b], 1+solve(x+1, y, bb-(1<<c)))
            if x < c-1 and b&(1<<(c-2)):
                m[x][y][b] = max(m[x][y][b], 1+solve(x+1, y, bb-(1<<(c-1))))
        return m[x][y][b]

    r, c = map(int, raw_input().split())
    a = [raw_input() for i in range(r)]
    
    m = [[[-1] * (1<<16) for i in range(r)] for j in range(c)]
    m1 = solve(0, 0, 0)
    a = map(lambda y: ''.join(map(lambda x: (x=='K' and '.') or x, y)), a)
    m = [[[-1] * (1<<16) for i in range(r)] for j in range(c)]
    m2 = solve(0, 0, 0)
    
    if m2 > m1:
        return "A"
    else:
        return "B"

for i in range(input()):
    print "Case #%d: %s" % (i+1, compute())

File: AMER Semifinal 2008/D.ml
(*  
 *  Problem: King
 *  Language: OCaml
 *  Author: KirarinSnow
 *  Usage: ocamlopt str.cmxa -o exec thisfile.ml && ./exec <input.in >output.out
 *  Comments: Adapted from reference solution.
 *)


open Scanf;;

let maxb = 1 lsl 16;;

let index y b = (y * maxb) + b;;

let rec solve a m x y b r c =
    if y = r then
      0
    else
      if x = c then
	solve a m 0 (y + 1) b r c
      else
	if m.(x).(index y b) <> (-1) then
	  m.(x).(index y b)
	else
	  let b2 = (b lsl 1) land ((1 lsl (c + 1)) - 1) in	     
	    if a.(y).[x] <> '.' then
	      m.(x).(index y b) <- solve a m (x + 1) y b2 r c
	    else
	      begin
		m.(x).(index y b) <-
		  solve a m (x + 1) y (b2 + 1) r c;
		
		if x != 0 && (b land 1) <> 0 then
		  m.(x).(index y b) <- max m.(x).(index y b)
		    (1 + (solve a m (x + 1) y (b2 - 2) r c));
	
		if x != 0 && (b land (1 lsl c)) <> 0 then
		  m.(x).(index y b) <- max m.(x).(index y b)
		    (1 + (solve a m (x + 1) y b2 r c));
		
		if (b land (1 lsl (c - 1))) <> 0 then
	            m.(x).(index y b) <- max m.(x).(index y b)
		      (1 + (solve a m (x + 1) y (b2 - (1 lsl c)) r c));
		
		if x < (c - 1) && (b land (1 lsl (c - 2))) <> 0 then
		    m.(x).(index y b) <- max m.(x).(index y b)
		      (1 + (solve a m (x + 1) y (b2 - (1 lsl (c - 1))) r c));
	      end;
	    m.(x).(index y b);;

let compute () =
  scanf "%d %d\n"
    (fun r c -> 
       let a = Array.create r "" in
	 for i = 0 to (r - 1) do
	   scanf "%s\n" (fun x -> a.(i) <- x)
	 done;
	 
	 let m = Array.make_matrix c (r * maxb) (-1) in
	 let m1 = solve a m 0 0 0 r c in
	   
         let a = Array.map
	   (fun l -> Str.global_replace (Str.regexp "K") "." l) a in
	   
	 let m = Array.make_matrix c (r * maxb)  (-1) in
	 let m2 = solve a m 0 0 0 r c in
	   
	   if m2 > m1 then "A" else "B"
    );;       


for i = 1 to read_int () do
  Printf.printf "Case #%d: %s\n" i (compute ());
done;;
